{% load format_tags %}
<a class="help" href="#help-motif"><img align="right" src="/static/images/help.png" /></a>
<h5> Motif information <span class="search_help">(de novo identified motifs for modules)</span>
</h5>
<div class="tabs">
<div class="main">
<p>
{% if motifs %}
There are {{motifs|length}} motifs predicted. {% if species.short_name == 'dvu' %}
<br>Click on the <a target="_blank" href="http://regpredict.lbl.gov/">RegPredict</a> links to explore the motif in <a target="_blank" href="http://regpredict.lbl.gov/">RegPredict</a>.{% endif %}</p>
<table id="motif">
  <caption> Motif Table ({{motifs|length}})</caption>
  <thead>
    <tr>
      <th>Motif Id</th>
      <th>e-value</th>
      <th>Consensus</th>
      <th>Motif Logo</th>
      {% if species.short_name == 'dvu' %}<th>RegPredict</th>{% endif %}
    </tr>
  </thead>
  <tbody>
    {% for m in motifs %}
  <tr>
    <td>{{m.id}}</td>
    <td>{{m.e_value|format_scientific}}</td>
    <td>{{m.consensus}}</td>
    <td>
      <div id="canvas_{{m.id}}"><img alt="Loader icon" src="{{STATIC_PREFIX}}images/ajax-loader.gif"></img></div>
    </td>
    {% if species.short_name == 'dvu' %}<td><a target="_blank" href="http://regpredict.lbl.gov/regpredict/com.lbl.regusw.App/App.html?taxonomyId={{species.ncbi_taxonomy_id}}&pssm={{m.pssm.as_string}}">RegPredict</a></td>{% endif %}
  </tr>
    {% endfor %}
  </tbody>
</table>
    {% for m in motifs %}
    <div class="gaggle-data ratios" style="display:none;">
        <span class="gaggle-name hidden">{{ species.short_name}} motif {{m.id}}</span>
        <span class="gaggle-species hidden">{{species.name}}</span>
        <span class="gaggle-size hidden">{{m.pssm|length}}x4</span>
        <div class="gaggle-matrix-tsv">
            {{ m.pssm.as_string }}
        </div>
    </div>
    {% endfor %}
{% else %}
<div class="error_message">
  <div class="main">
	<div class="ui-widget">
    	<div class="ui-state-highlight ui-corner-all" style="padding: 0 .7em;">
    		<p><span class="ui-icon ui-icon-info" style="float:left;margin-right:.3em;"></span>
			<strong> Warning:</strong> No Motifs were found!</p>
		</div>
	</div>
</div>

</div>
{% endif %}
<div style="display:none">
	<div id='help-motif' class="help" style='padding:10px; background:#fff; text-align:left;'>
    <h5>Motif Help</h5>
    	<p>Transcription factor binding motifs help to elucidate regulatory mechanism. cMonkey integrates powerful de novo motif detection to identify conditionally co-regulated sets of genes. De novo predicted motifs for each module are listed in the module page as motif logo images along with associated prediction statistics (e-values). The main module page also shows the location of these motifs within the upstream sequences of the module member genes.</p>
    	<p> Motifs of interest can be broadcasted to RegPredict (currently only available for Desulfovibrio vulgaris Hildenborough) in order to compare conservation in similar species. This integrated motif prediction and comparative analysis provides an additional checkpoint for regulatory motif prediction confidence.</p>
		 <p><b> Motif e-value:</b> cMonkey tries to identify two motifs per modules in the upstream sequences of the module member genes.
    	Motif e-value is an indicative of the motif co-occurences between the members of the module.Smaller e-values are indicative of significant sequence motifs. Our experience showed that e-values smaller than 10 are generally indicative of significant motifs.</p>
	</div>
 </div>
 
</div>
</div>
