{% load staticfiles %}
<div id="content">
  <div class="sub-page">
    <h5> Gene Page Help</h5>
<br>
<div id="network" class="help" style='padding:10px; background:#fff; text-align:left;'>
<h2 id="network">Network Tab</h2>
<p> If the gene is associated with a module(s), its connection to given modules along with other members of that module are shown as network by using CytoscapeWeb. In this view, each green colored circular nodes represent module member genes, purple colored diamonds represent module motifs and red triangles represent regulators. Each node is connected to module (Bicluster) via edges. This representation provides quick overview of all genes, regulators and motifs for modules. It also allows one to see shared genes/motifs/regulators among diferent modules. <br><br>
Network representation is interactive. You can zoom in/out and move nodes/edges around. Clicking on a node will open up a window to give more details. For genes, Locus tag, organism, genomic coordinates, NCBI gene ID, whether it is transcription factor or not and any associated functional information will be shown. For regulators, number of modules are shown in addition to gene details. For motifs, e-value, consensus sequence and sequence logo will be shown. For modules, expression profile plot, motif information, functional associations and motif locations for each member of the module will be shown.<br>
You can pin information boxes by using button in the box title and open up additional ones on the same screen for comparative analysis.<br>
</div>
<div id="regulation" class="help" style='padding:10px; background:#fff; text-align:left;'>
<h2 id="regulation">Regulation Tab</h2>   
<p> Regulation tab for each gene includes regulatory influences such as environmental factors or transcription factors or their combinations identified by regulatory network inference algorithms.</p>
<p> If the gene is a member of a module, regulators influencing that module are also considered to regulate the gene. Regulators table list total number of regulatory influences, regulators, modules and type of the influence.</p>
<p> You can see description of the regulator inside the tooltip when you mouseover. In certain cases the regulatory influence is predicted to be the result of the combination of two influences. These are indicated as combiner in the column labeled "Operator".</p>
<p> For transcription factors, an additional table next to regulator table will be show. This table show modules that are influenced by the transcription factor. </br>
</div>
<div id="motif" class="help" style='padding:10px; background:#fff; text-align:left;'>
  <h2 id="motifs">Motifs Tab</h2>   
<p> Network inference algorithm uses de novo motif prediction for assigning genes to modules. If there are any motifs identified in the upstream region of a gene, the motif will be shown here. For each motif sequence logo, consensus and e-value will be shown.<br>
 </div>
<div id="functions" class="help" style='padding:10px; background:#fff; text-align:left;'>
  <h2 id="functions">Functions Tab</h2>  
<p> Identification of functional enrichment for the module members is important in associating predicted motifs and regulatory influences with pathways. As described above, the network inference pipeline includes a functional enrichment module by which hypergeometric p-values are used to identify over representation of functional ontology terms among module members.</p>
<p> Network Portal presents functional ontologies from KEGG, GO, TIGRFAM, and COG as separate tables that include function name, type, corrected and uncorrected hypergeometric p-values, and the number of genes assigned to this category out of total number of genes in the module.<br>
</div>
<div id="regulon-members" class="help" style='padding:10px; background:#fff; text-align:left;'>
  <h2 id="regulon-members">Module Members Tab</h2> 
 <p> Identity of gene members in a module may help to identify potential interactions between different functional modules. Therefore, neighbor genes that share the same module(s) with gene under consideration are shown here. For each memebr, gene name, description and modules that contain it are listed. <br>
</div>
<div id="help" class="help" style='padding:10px; background:#fff; text-align:left;'>
    <h2 id="help">Help Tab</h2> 
<p> This help page. More general help can be accessed by clicking help menu in the main navigation bar.</p>
</div>
<div id="social" class="help" style='padding:10px; background:#fff; text-align:left;'>
     <h2 id="comments">Social Tab</h2>
<p> Network Portal is designed to promote collaboration through social interactions. Therefore interested researchers can share information, questions and updates for a particular gene.</p>
<p> Users can use their Disqus, Facebook, Twitter or Google accounts to connect to this page (We recommend Google). Each module and gene page includes comments tab that lists history of the interactions for that gene. You can browse the history, make updates, raise questions and share these activities with social web.</p>
<p> In the next releases of the network portal, we are planning to create personal space for each user where you can share you space that contains all the analysis steps you did along with relevant information. 
</p>
</div>
<h2 id="circvis">CircVis</h2>
Our circular module explorer is adapted from <a href="http://code.google.com/p/visquick/">visquick</a> originally developed by Dick Kreisberg of <a href="http://shmulevich.systemsbiology.net/">Ilya Shmulevich lab</a> at ISB for <a href="http://cancergenome.nih.gov/">The Cancer Genome Atlas</a>. We use simplified version of visquick to display distribution of module members and their interactions across the genome. This view provides summary of regulation information for a gene. The main components are;
<img src="{% static "images/help_images/circvis-help.png" %}" align="right"><br>
  <ul>
    <li> 1. All genomic elements for the organism are represented as a circle and each element is separated by black tick marks. In this example chromosome and pDV represent main chromosome and plasmid for D. vulgaris Hildenborough, respectively.</li>
    <li> 2. Source gene</li>
    <li> 3. Target genes (other module members)</li>
    <li> 4. Interactions between source and target genes for a particular module</li>
    <li> 5. Module(s) that source gene and target genes belong to</li>
    <li> 6. Visualisation legend</li>
  </ul>
</div>
</div>
