{% extends "base2.html" %}
{% load staticfiles %}
{% block title %}Network Portal - Help {% endblock %}
{% block cssspecial %}
<link rel="stylesheet" href="http://code.jquery.com/ui/1.9.2/themes/cupertino/jquery-ui.css"
      xmlns="http://www.w3.org/1999/html" xmlns="http://www.w3.org/1999/html"/>
<link href="{% static "javascripts/jquery.qtip.min.css" %}" rel="stylesheet" type="text/css" />
<link rel="stylesheet" href="{{STATIC_PREFIX}}stylesheets/colorbox.css" />
{% endblock %}
{% block jsspecial %}<script type="text/javascript" src="{% static "cytoscapeweb/js/AC_OETags.min.js" %}"></script>
    <script type="text/javascript" src="{% static "cytoscapeweb/js/cytoscapeweb.min.js" %}"></script>
    <script type="text/javascript" src="{% static "javascripts/wwu_logo/isblogo.js" %}"></script>
    <script type="text/javascript" src="{% static "javascripts/network_helpers.js" %}"></script>
    <script type="text/javascript" src="{% static "javascripts/jquery.qtip.min.js" %}"></script>
    <script type="text/javascript" src="http://cdn.wijmo.com/jquery.wijmo-open.1.5.0.min.js"></script>
    <script type="text/javascript" src="{% static "javascripts/protovis.min.js" %}"></script>
    <script type="text/javascript" src="{% static "javascripts/visquick-min/vq.min.js" %}"></script>
    <script type="text/javascript" src="{% static "javascripts/visquick-min/vq.circvis.min.js" %}"></script>
    <script type="text/javascript" src="{% static "javascripts/visquick_helpers.js" %}"></script>
    <script type="text/javascript" src="http://www.ncbi.nlm.nih.gov/core/extjs/ext-3.4.0/adapter/ext/ext-base.js"></script>
    <script type="text/javascript" src="http://www.ncbi.nlm.nih.gov/core/extjs/ext-3.4.0/ext-all.js"></script>
    <script type='text/javascript' src='{{STATIC_PREFIX}}javascripts/jquery.colorbox.js'></script>
    <script type="text/javascript">window.sv_base_url = "/sviewer/"</script>
    <script type="text/javascript">
      $(document).ready(function()
      {
      // We'll target all AREA elements with alt tags (Don't target the map element!!!)
      $('area[alt]').qtip(
      {
      content: {
      attr: 'alt' // Use the ALT attribute of the area map for the content
      },
      });
      $(".figure1").colorbox({rel:'figure1'});
      });
    </script>
    <title>ScreenFlow video</title>
    <meta http-equiv="content-type" content="text/html;charset=iso-8859-1" />
    <meta http-equiv="Content-Style-Type" content="text/css" />
    <meta http-equiv="Content-Script-Type" content="text/Javascript" />
    <script type="text/javascript" src="{% static "Network_Portal_Screencast_01/com/js/swfobject/swfobject.js" %}"></script>
    <script type="text/javascript">
      var flashvars = {};
      var params = {};
      
      params.bgcolor = "#000000";
      params.allowscriptaccess = "always";
      
      flashvars.videoPath = "{% static "Network_Portal_Screencast_01/movie.f4v" %}";
      flashvars.posterPath = "{% static "Network_Portal_Screencast_01/com/poster/myPosterFrame.jpg" %}";
      flashvars.skinPath = "{% static "Network_Portal_Screencast_01/com/skin/skin.swf" %}";
      
      var stageW = 572;
      var stageH = 564;
      
      var attributes = {};
      attributes.id = "flashcontent";

      swfobject.embedSWF("{% static "Network_Portal_Screencast_01/player.swf" %}",
        "flashcontent", stageW, stageH, "9.0.0", false, flashvars, params, attributes);
    </script>
		<script>
			$(function() {
				$( "#accordion-portal" ).accordion({
				collapsible: true,
				heightStyle: "content"
				});
			});
		</script>

        <script>
			$(function() {
				$( "#accordion" ).accordion({
				collapsible: true,
				heightStyle: "content"
				});
			});
		</script>
		<script>
		$(function() {
			$( "#accordion2" ).accordion({
				collapsible: true,
				heightStyle: "content" 
			});
		});
		</script>

     <script>
        $(function() {
            $( "#tabs" ).tabs();
        });
    </script>
    
     <!-- These are required for loading NCBI sequence viewer in gene page -->
    <!--<script type="text/javascript" src="http://www.ncbi.nlm.nih.gov/projects/sviewer/js/main.js"/>-->
    {% endblock %}
    {% block content %}    
    <div class="sub-page">
	<h5> Network Portal Help</h5>
	<h4 align="right"> <a href="https://plus.google.com/111789053530927127952/videos"><img src="{{ STATIC_PREFIX }}images/Google_plus_32.png" align="center"></a> Please visit <a href="https://plus.google.com/111789053530927127952/videos">screencast tutorials</a> page to quickly learn how to use Gaggle and Gaggle workspace</h4><p>
		<div id="tabs">
			<ul>
				<li><a href="#tabs-1">Discover Network Portal</a></li>
				<li><a href="#tabs-2">What is Gaggle?</a></li>
				<li><a href="#tabs-3">Gaggle Workspace</a></li>
				<li><a href="#tabs-4">Screencast</a></li>
				<li><a href="#tabs-5">Get Involved</a></li>
			</ul>
			<div id="tabs-1"> <h5> Discover Network Portal</h5>
				<div id="accordion-portal">
					<h4>What is Network Portal</h4>
					<div><p>The Network Portal is a database of gene transcription regulatory networks and enables exploration, annotation and comparative analysis.</p>
					<p> Deciphering the complexity of biological systems requires a systems-level view of regulatory players and their interactions. However, inference of these complex interactions is challenging, and further, visualizing and analyzing these interactions often requires advanced computational expertise, tools and resources.</p> 
					<p>Network Portal provides analysis and visualization tools for selected gene regulatory networks to aid researchers in biological discovery and hypothesis development. It provides a user-friendly interface, extensive search capabilities and a number of visualization tools. We are also constantly adding regulatory networks for new species. Increasing the number of network models will also enable comparative network biology. We are also developing tools for comparing network models and these will be integrated into the portal as soon as they are available.</p>
					<p> The Network Portal also acts as a gateway for the Gaggle Workspace. Gaggle Workspace makes the life of researchers easier by taking over the burden of repetitive analysis tasks. Integration with Network Portal provides full access to automated analysis of network models and associated biological information.</p>
					</div>
					<h4>Navigating Gene Pages</h4>
					<div>
						<div id="tab-help">
							<h2 id="network">Network Tab</h2>
							<p>If the gene is associated with a regulon(s), its connection to given regulons along with other members of that regulon are shown as a network using CytoscapeWeb. In this view, green circular nodes represent regulon member genes, purple diamonds represent regulon motifs and red triangles represent regulators. Each node is connected to regulon(s) (Biclusters) via edges. This representation provides a quick overview of all genes, regulators and motifs for regulons. It also allows one to see shared genes/motifs/regulators among different regulons. Network representation is interactive. You can zoom in/out and move nodes/edges around. Clicking on a node will open up a window to give more details.</br> The information shown for genes includes the locus tag, organism, genomic coordinates, NCBI gene ID, whether it is transcription factor and any associated functional information. For regulators, the number of regulons is shown in addition to gene details. For motifs, e-value, consensus sequence and sequence logo will be shown. For regulons, expression profile plot, motif information, functional associations and motif locations for each member of the regulon will be shown. You can pin information boxes by using the button in the box title and open up additional ones on the same screen for comparative analysis.</p>								
							<h2 id="regulation">Regulation Tab</h2>   
								<p> The regulation tab for each gene includes regulatory influences such as environmental factors or transcription factors or their combinations identified by regulatory network inference algorithms. If the gene is a member of a regulon, regulators influencing that regulon are also considered to regulate the gene. The regulators table lists the total number of regulatory influences, regulators, regulons and type of the influence. You can see a description of the regulator inside the tooltip when you mouseover. In certain cases the regulatory influence is predicted to be the result of the combination of two influences. These are indicated as a combiner in the column labeled "Operator". For transcription factors, an additional table next to regulator table will be shown. This table shows regulons that are influenced by the transcription factor. </br>
									<h2 id="motifs">Motifs Tab</h2>   
									<p> Network inference algorithm uses de novo motif prediction for assigning genes to regulons. If there are any motifs identified in the upstream region of a gene, the motif will be shown here. For each motif sequence logo, consensus and e-value will be shown.<br>
									<h2 id="functions">Functions Tab</h2>  
									<p> Functional annotations for a given gene might be important to connect regulatory information with physiology of the cell. Therefore, functional annotations for each gene collected from KEGG, Gene Ontology (GO), Cluster of Orthologous Groups (COG) and TIGR Roles will be shown here. The function name is linked to its explanation and more details.<br>
									<h2 id="regulon-members">Regulon Members Tab</h2> 
									<p> Identity of gene members in a regulon may help to identify potential interactions between different functional modules. Therefore, neighbor genes that share the same regulon(s) with gene under consideration are shown here. For each member, gene name, description and regulons that contain it are listed. <br>
									<h2 id="help">Help Tab</h2> 
									<p> More general help can be accessed by clicking the help menu in the main navigation bar.</p>
									<h2 id="comments">Social Tab</h2>
									<p> Network Portal is designed to promote collaboration through social interactions. Therefore, interested researchers can share information, questions and updates for a particular gene. Users can obtain login information by registering for our website. Alternatively, they can use their Facebook or Google accounts to connect to this page. Each regulon and gene page includes a comments tab that lists the history of interactions for that gene. You can browse the history, make updates, raise questions and share these activities within the social web. In future releases of the network portal, we are planning to create a personal space for each user that contains all the analysis steps performed by the user along with relevant information. It will be possible to share these personal spaces with other users.</p>
									<h2 id="circvis">CircVis</h2>
														Our circular regulon explorer is adapted from <a href="http://code.google.com/p/visquick/">visquick</a>, which was originally developed by Dick Kreisberg of <a href="http://shmulevich.systemsbiology.net/">Ilya Shmulevich lab</a> at ISB for <a href="http://cancergenome.nih.gov/">The Cancer Genome Atlas</a>. We use a simplified version of visquick to display distribution of regulon members and their interactions across the genome. This view provides summary of regulation information for a gene. The main components are;
														<img src="{% static "images/help_images/circvis-help.png" %}" align="right"><br>
															<ul>
																<li> 1. All genomic elements for the organism are represented as a circle and each element is separated by black tick marks. In this example "chromosome" and "pDV" represent the main chromosome and plasmid for D. vulgaris Hildenborough.</li>
																<li> 2. Source gene</li>
																<li> 3. Target genes (other regulon members)</li>
																<li> 4. Interactions between source and target genes for a particular regulon</li>
																<li> 5. Regulon(s) that the source gene and target genes belong to</li>
																<li> 6. Visualization legend</li>
															</ul>

						</div>
					</div>
					<h4>Navigating Module Pages</h4>
					<div>
						<div id="tab-navigate">
					<h2> What is a module?</h2>
			<p>Regulatory units (modules) in the Network Portal are based on the network inference algorithm used. For the current version, modules are based on <a target="blank" href="http://baliga.systemsbiology.net/drupal/content/new-cmonkey-r-package-and-code">cMonkey</a> modules and <a target = "blank" href="http://baliga.systemsbiology.net/drupal/content/inferelator">Inferelator</a> regulatory influences on these modules. More specifically, module refers to set of genes that are conditionally co-regulated under subset of the conditions. Identification of modules integrates co-expression, de-novo motif identification, and other functional associations such as operon information and protein-protein interactions.</p>
			<h2>Module Overview</h2>
			<a class="figure1" href="/static/images/module-figure.png" title="An Example summary view of the module"><img height=20% width=20% align="left" src="/static/images/module-figure.png"></a>
			<p> The landing module page shows quick summary info including co-expression profiles, de-novo identified motifs, and transcription factors and/or environmental factors as regulatory influences. It also includes module residual, motif e-values, conditions and links to other resources such as NCBI and Microbesonline. 
		  . If a transcription factor is included in the manually curated <a target="blank" href="http://regprecise.lbl.gov/RegPrecise/">RegPrecise</a> database, further information from RegPrecise is shown, allowing users to perform comparative analysis.</p>
		  	<h2> Expression Profiles</h2>
		  	<p>Expression profiles is a plot of the expression ratios (log10) of the module's genes, over all subset of the conditions included in the module.
		  	The X-axis represent conditions and the Y-axis represents log10 expression ratios. Each gene is plotted as line plot with different colors. 
		  	Colored legend for the lines are presented under the plot. This plot is dynamic. Clicking on the gene names in the legend will show/hide the plot for that particular gene. A tooltip will show expression ratio information if you mouseover the lines in the plot.</p>
		  	<h2> Motif Locations</h2>
		  	<p>Location of the Identified motifs for the module in the upstream regions of the member genes are shown under the expression profiles plot. This plot shows the diagram of the upstream positions of the motifs, colored red and green for motifs #1, and 2, respectively. Intensity of the color is proportional to the significance of the occurence of that motif at a given location.	Motifs on the forward and reverse strand are represented over and under the line respectively.</p>	
		  	</div>
		<div id="network-info" style='padding:10px; background:#fff; text-align:left;'>
		 	<h2>Network</h2>
		 	<p>A network view of the module is created using cytoscapeWeb and enables dynamic, interactive exploration of the module properties. In this view, module member genes, motifs, and regulatory influences are represented as peripheral nodes connected to core module node via edges. Module members are green circles, regulators are red triangles and motifs are blue diamonds. Selection of a node gives access to detailed information in a pop-up window, which allows dragging and pinning to compare multiple selections. Selecting module members will show information about the selected gene such as name, species and fucntions. Motif selection will show motif logo image and e-values. Bicluster selction will show expression profile and summary statistics for the module.</p>
		 	<img alt="Gene" src="{{ STATIC_PREFIX }}images/network_gene.png" style="background:none;border:0px;position:relative; bottom:-10px;">Module member <img alt="Regulator" src="{{ STATIC_PREFIX }}images/network_regulator.png" style="background:none;border:0px;position:relative; bottom:-10px;">Regulator  <img alt="Motif" src="{{ STATIC_PREFIX }}images/network_motif.png" style="background:none;border:0px;position:relative; bottom:-10px;">Motif
		</div>
		<div id="regulator-info" style='padding:10px; background:#fff; text-align:left;'>
		 	<h2 id="regulators"> Regulators</h2>
		 	<p>For each module, single or AND logic connected regulatory influences are listed under the regulators tab. These regulatory influences are identified by Inferelator. Table shows name of the regulator and its type. <b>tf:</b> Transcription factor, <b>ef:</b> Environmental factor and <b> combiner:</b>Combinatorial influence of a tf or an ef through logic gate. Tabel is sortable by clicking on the arrows next to column headers.</p>
		</div>
		<div id="motif-info" style='padding:10px; background:#fff; text-align:left;'>
		 	<h2>Motifs</h2>
    		<p>Transcription factor binding motifs help to elucidate regulatory mechanism. cMonkey integrates powerful de novo motif detection to identify conditionally co-regulated sets of genes. De novo predicted motifs for each module are listed in the module page as motif logo images along with associated prediction statistics (e-values). The main module page also shows the location of these motifs within the upstream sequences of the module member genes.</p>
    		<p> Motifs of interest can be broadcasted to <a target="blank" href="http://regpredict.lbl.gov/regpredict/">RegPredict</a> (currently only available for Desulfovibrio vulgaris Hildenborough) in order to compare conservation in similar species. This integrated motif prediction and comparative analysis provides an additional checkpoint for regulatory motif prediction confidence.</p>
    	</div>
    	<div id="functions-info" style='padding:10px; background:#fff; text-align:left;'>
    		<h2>Functions</h2>
    		<p>Biological networks contain sets of regulatory units called functional modules that together play a role in regulation of specific functional processes. Connections between different modules in the network can help identify regulatory relationships such as hierarchy and epistasis. In addition, associating functions with modules enables putative assignment of functions to hypothetical genes.  It is therefore essential to identify functional enrichment of modules within the regulatory network.</p>
    		<p> Functional annotations from single sources are often either not available or not complete. Therefore, we integrated KEGG pathway, Gene Ontology, TIGRFam and COG information as references for functional enrichment analysis.</p>
    		<p> We use hypergeometric p-values to identify significant overlaps between co-regulated module members and genes assigned to a particular functional annotation category. P-values are corrected for multiple comparisons by using Benjamini-Hochberg correction and filtered for p-values â‰¤ 0.05.</p>
    		<p>Network Portal presents functional ontologies from KEGG, GO, TIGRFAM, and COG as separate tables that include function name, type, corrected and uncorrected hypergeometric p-values, and the number of genes assigned to this category out of total number of genes in the module.</p>
    	</div>
    	<div id="gene-info" style='padding:10px; background:#fff; text-align:left;'>
		 	<h2>Genes</h2>
    		<p> Gene member table shows all the genes included in the module. Listed attributes are;
    		<ol>
    			<li> Name: Gene name or Locus tag</li>
    			<li> Common Name: Gene short name</li>
    			<li> Type: Type of the feature, usually CDS.</li>
    			<li> Gene ID: Link to NCBI Gene ID</li>
    			<li> Chromosome: Chromosome name from annotation file</li>
    			<li> Start/End:Feature start and end coordinates</li>
    			<li> Strand: strand of the gene</li>
    			<li> Description: Description of the gene from annotation file</li>
    			<li> TF: If the gene is a Transcription Factor or not.</li>
    		</ol>
    		<p>If you are browsing the Network Portal by using Gaggle/Firegoose, firegoose plugin will capture the NameList of the gene members. Captured names can be saved into your Workspace by clicking on <b> "Capture"</b> in the firegoose toolbar or can be directly sent other desktop and web resources by using <b>"Broadcast"</b> option.
        	</div>

    	<div id="social-info" style='padding:10px; background:#fff; text-align:left;'>
    		<h2>Social</h2>
			<p>You can start a conversation about this module or join the existing discussion by adding your comments. In order to be able to add your comments you need to sign in by using any of the following services;Disqus, Google, Facebook or Twitter. For full compatibility with other network portal features, we recommend using your Google ID.</p>  
		 </div>
		 <div id="definitions-info" style='padding:10px; background:#fff; text-align:left;'>
		 	<h2> Definitions</h2>
		 	<p><b>Residual:</b> is a measure of bicluster quality. Mean bicluster residual is smaller when the expression profile of the genes in the module
    	is "tighter". So smaller residuals are usually indicative of better bicluster quality.</p>
    		<p><b>Expression Profile:</b> is a preview of the expression profiles of all the genes under subset of conditions included in the module.
    	Tighter expression profiles are usually indicative of better bicluster quality.</p>
    		<p><b> Motif e-value:</b> cMonkey tries to identify two motifs per modules in the upstream sequences of the module member genes.
    	Motif e-value is an indicative of the motif co-occurences between the members of the module.Smaller e-values are indicative of significant sequence motifs. Our experience showed that e-values smaller than 10 are generally indicative of significant motifs.</p>
    		<p><b> Genes:</b> Number of genes included in the module.</p>
    		<p><b> Functions:</b> We identify functional enrichment of each module by camparing to different functional categories such as KEGG, COG, GO etc. by using hypergeometric function. If the module is significantly enriched for any of the functions, this column will list few of the these functions as an overview. 
    	Full list of functions is available upon visiting the module page under the Functions tab.</p>

					</div>
					</div><!-- div for navigating module pages -->		
					
				</div>
			</div>
			<div id="tabs-2"> <h5> What is Gaggle?</h5>
				<div id="accordion" class="">
					<h4> Overview</h4>
					<div><p>The practice of systems biology depends upon many software tools, operating on many kinds of data from many different sources. Each of these tools typically excels at one (or a few) types of analysis with one (or a few) types of data. A crucial challenge, therefore, is to combine the capabilities of these and other, forthcoming tools to create a data exploration and analysis environment which can do justice to the variety and complexity of systems biology.</p>
						<p>The Gaggle is a simple, open-source Java software environment which solves this problem. Guided by the classic software engineering strategy of separation of concerns and a policy of semantic flexibility, it combines existing popular programs and web resources into a user-friendly, rich, and easily extended environment in which to do systems biology.</p>

						<p>We currently support a number of geese -- our name for any open source software which is adapted to run in the gaggle. This adaptation is generally only a small amount of programming work. Once gaggled the program can broadcast and receive any of a small number of data types which together constitute an adequate basis for exploratory analysis in systems biology. These data types include:</br></p>
						<ol>
							<li><strong>Name list</strong> (i.e., these genes are interesting)</li>
							<li><strong>Bicluster:</strong> Name list combined with a condition list (i.e., these genes are interesting in these conditions)</li>
							<li><strong>Tuple</strong> (replaces HashMap): a collection of name/value pairs</li>
							<li><strong>Matrix:</strong> rows and columns, each named, containing numerical data</li>
							<li><strong>Network:</strong> a collection of nodes and edges, with arbitrary tuples associated with each</li>
						</ol>

					</div>
					
					<h4> Current Geese</h4>
					<div class="post-tab">
						<img align="right"src="http://gaggle.systemsbiology.net/docs/geese/schematic.png" />
						The Boss acts as a hub that routes messages from one goose to another.
						<ul>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/boss.php">Gaggle Boss</a> <span class="ui-state-error"> Workflow Enabled</span></li>
						</ul>
						A database or software program that is capable of communicating with the Gaggle framework
							is called a <strong>Goose</strong>. Here is a (partial) list of Geese:
						<ul>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/anno.php">Annotation</a>. Search for functional annotations.</li>
							<li><a href="http://www.sysbio.org/research/bsi/bioanalytics/bioinformatics.stm">Bioinformatics Resource Manager</a>. A general purpose data management, analysis and integration environment for systems biology.</li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/cytoscape.php">Cytoscape</a>. Network visualization.<span class="ui-state-error"> Workflow Enabled</span></li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/dmv.php">DMV</a>. Datamatrix Viewer.</li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/firegoose">Firegoose</a>. Connect the Gaggle to the Web.<span class="ui-state-error"> Workflow Enabled</span></li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/genomebrowser">Genome Browser</a>. View data in genomic context.</li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/matgoose/">MatGoose</a>. Exchange data between MatLab and the Gaggle.</li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/tigrmev.php">MeV</a>. MultiExperiment Viewer. Cluster and visualize microarray data.<span class="ui-state-error"> Workflow Enabled</span></li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/r">R Goose</a>. Exchange data between the R statistical environment and the Gaggle.</li>
							<li><a href="http://gaggle.systemsbiology.net/docs/geese/translator/">Translator</a>. Translate identifiers between naming systems or (by orthology) between organisms.</li>
							<li><a href="http://gaggle.systemsbiology.net/wiki/doku.php?id=json_script_goose">Script Goose</a>. A goose with an embedded HTTP server and command-line client, allowing programs written in any language to communicate with the Gaggle.</li>
							</ul>
						<p>Gaggle is designed to make it easy to connect programs and databases with a minimum of effort. There are several <a href="http://gaggle.systemsbiology.net/docs/geese/thirdparty/">third party Gaggle-connected applications</a>.</p>
						</div>
						<h4> Gaggle Development Principles</h4>
						<div><style type="text/css">
							A:link {color:#3300ff}
							A:visited {text-decoration: none}
							A:active {text-decoration: none}
							A:hover {text-decoration: underline; color: red;}
							</style>
							<h4>Core components</h4>

						<ul>
							<li> The core components of the Gaggle are the <strong>Boss</strong> and the <strong>Goose</strong> interfaces. Because all Geese depend on these interfaces, changes to them require changes to all Geese. Therefore, changes to these interfaces must be carefully controlled to maintain compatibility.

								<ul>
									<li> The <strong>Boss</strong> and <strong>Goose</strong> interfaces should be revised rarely and carefully.</li>
									<li> The right to change core components is reserved by the Baliga Lab at ISB.</li>
									<li> <a href="/mantis" title="requested_features">Feature requests</a> for core components will be carefully considered.</li>
								</ul>
							</li>
						</ul>

						<h4>Loose Coupling</h4>
						<ul>
							<li> The Gaggle is designed as a system of loosely coupled components:
								<ul>
									<li> Almost any desired functionality can be implemented as a Goose.
									</li>
								</ul>
							</li>
						</ul>
						<h4>Data Types</h4>
						<ul>
							<li> Existing Data types:
								<ul>
									<li> name list
									</li>
									<li> tuple
									</li>
									<li> matrix
									</li>
									<li> network
									</li>
									<li> cluster
									</li>
								</ul>
							</li>
						</ul>
						<ul>
							<li> A few general data types are preferred over special-purpose data types.
							</li>
							<li> Almost all biological data can be represented as one of the existing types.
							</li>
							<li> New data types should be introduced only when necessary.
							</li>
						</ul>

					<h4>Semantic Flexibility</h4>
					<ul>
						<li> Since the Gaggle is a merely a conduit for transferring data between applications, it does not need to assign specific meaning or roles to the data.
						</li>
						<li> Interpreting the data is left up to the user and the applications.
						</li>
					</ul>

					<h4>Open Development</h4>
					<ul>
						<li> The Gaggle is open source.
						</li>
						<li> The <a href="http://gaggle.systemsbiology.net/svn/gaggle/gaggle/trunk" class="urlextern" target="_new" title="http://gaggle.systemsbiology.net/svn/gaggle/gaggle/trunk"  rel="nofollow">source repository</a> for the core components of the Gaggle will be maintained at ISB.
						</li>
						<li> Permission to commit to the source repository will be granted on a case-by-case basis. Reading the source repository is open to all.
						</li>
						<li> Contributions of code, especially new Geese, are welcome.
						</li>
					</ul>

					<h4>The Gaggle Paper</h4>
					<p>These principles are more fully explained in the
					<a href="http://www.biomedcentral.com/1471-2105/7/176" target="_blank">Gaggle paper</a>.</p>
					</div>
				</div>
			</div> 
			   
			<div id="tabs-3"> <h5> Gaggle Workspace</h5>
				<div id="accordion2" class="">
					<h4> Workspace Concept</h4>
                    <div>
                        <p>
                            Gaggle Workspace offers a comprehensive data analysis environment allowing users to:
                        </p>
                        <ol style="margin-left: 50px">
                            <li>Capture and organize data from websites and local files;</li>
                            <li>Manipulate data from the Workspace and via automatically opened applications;</li>
                            <li>Create workflows automatically or manually which include multiple applications, websites, and datasets;</li>
                            <li>Save the state of an analysis session across all open websites and desktop applications and reload the state later from any machine;</li>
                            <li>Automatically generate summary reports;</li>
                            <li>Keep a log of recent activities.</li>
                        </ol>
                        <h4>FAQ</h4>
                        <p>
                            <u>1. What is the Dataspace tab? How can I upload data to it?</u><br />
                            The Dataspace is a central place to store and manipulate the data.  There are two types of data.
                            First, the system generated data, which is in the "Network Portal Files" table of the Dataspace tab.
                            Second, user captured data, which is displayed in the "User Files" table.
                            You can use the "Capture" button of the Firegoose to capture data from any web page (See <a href="http://www.youtube.com/watch?v=Z93VKagR6os">this</a> video).
                            You can also upload your file to the workspace (<a href="http://www.youtube.com/watch?v=XrZMnaTLW0A">this</a> video explains how).
                            You can save these data so you can reuse them later.
                        </p>
                        <p>
                            <u>2. What can I do with the data in the Dataspace? What applications are supported by the Gaggle Workspace?</u> <br />
                            You can open the data by selecting "Open" from the dropdown list on each row of the tables.
                            Gaggle will try to automatically start the Boss and the goose associated with the data.
                            For example, we will automatically start MeV if the file is a "MeV Analysis File".
                            If we cannot decide which goose to open, a dialog will pop up and you can select which one to use to
                            open the data. <a href="http://www.youtube.com/watch?v=K9ojLoqbRSI">This video</a> explains how to open data from the workspace.
                            Once the geese are open, you can do analysis and broadcast data between them.
                            For now we support three geese: Firegoose, Cytoscape, and MeV. Other geese will be added later.
                        </p>
                        <p>
                            <u>3. What is the Workflow tab? How can I build a workflow?</u><br />
                            The Gaggle Workspace supports the concept of workflow. It is very often that biologists need to perform repetitive analysis tasks
                            where the same/different inputs are analyzed by using the same/different tools multiple times. In addition, a number
                            of analyses tools may need to be used for analyzing the same set of inputs.
                            The workflow allows automation of these repetitive tasks by linking inputs and outputs for various desktop and web based analysis tools.<br />

                            There are two ways you can create a workflow. First, the system will automatically record the
                            geese you opened, and the source and target of each broadcast, and display it as a workflow
                            on the "Workflow" page.
                            You can also manually create workflows. To do this, you drag a rectangle with goose name
                            under the "Modules" column, and drop it on to the Workflow Canvas. A box will show up.
                            It contains the execution path of the goose (you can change it if you know the installation path).
                            Tutorial video can be found <a href="http://www.youtube.com/watch?v=EwEC0WL9BKg">here</a>.

                            More details about the workflow can be found <a href="#workflowconceptbookmark">here</a>.
                        </p>
                        <p>
                            <u>4. What is save/load states? How can I use this feature?</u><br />
                            Save/load state is a very useful feature. You can go to the "Saved States" tab and click the "Save State" button.
                            What it will do is it will contact each geese and tell them to save their current state.
                            The state files will be saved on the server so it takes some time. Once it is done, you will
                            see an notification and a new row appear in the saved states table.  You can reload the state from
                            any machine later on.  Note that currently only Firegoose, Cytoscape, and MeV support saving state.
                            We are actively working to enable more geese.  <a href="http://www.youtube.com/watch?v=qiB7rs_2wW4">This video</a> explains save/load session states.
                        </p>
                        <p>
                            <u>5. What are session reports?</u><br />
                            Session reports are automatically generated for workflows. A report consists of
                            snapshots of each applications of the workflow after the workflow is finished.  <a href="http://www.youtube.com/watch?v=Rk3Bv2vochA">This video</a>
                            explains session reports.
                        </p>
                        <p>
                            A series of video tutorials can be found <a href="https://plus.google.com/u/1/b/111789053530927127952/111789053530927127952/videos?cfem=1">here</a>.
                        </p>
                    </div>
                    <h4> Workflow Concept<a name="workflowconceptbookmark"></a></h4>
					<div>
						<p><img align="right" src="{% static "images/workflow_concept.png" %}">Researchers often come across repetitive analysis tasks where the same/different inputs are analyzed by using the same/different tools multiple times. In addition, a number of analyses tools may need to be used for analyzing the same set of inputs. The workflow allows automation of these repetitive tasks by linking inputs and outputs for various desktop and web based analysis tools. </p>
						<p>The Workflow is developed based on our extensive experience with Gaggle in exchanging data between independently developed software tools and databases to enable interactive exploration of systems biology data. Exporting the "boss" and "goose" concept into the workflow, each goose becomes a module in the workflow. These modules can be added into a workflow sequentially or in parallel and can accept output from other modules in the workflow. Once the workflow is built, it can be saved into a personal workspace and can be called out anytime to invoke the same analysis steps.</p>
					</div>
					<h4> Case 1 - Discovering Detailed Regulon Information</h4>
					<div> <img align="right" src="{% static "images/workflow_case-1.png" %}">
					<p> This exercise starts with a regulon in Network Portal. We will use a very simple workflow idea that includes broadcasting data via Firegoose to multiple web resources simultaneously. We will broadcast a Namelist of regulon members to DAVID for functional enrichment analysis and to Entrez to access NCBI tools. We will also broadcast motif PSSM information to STAMP in order to identify matches to known motifs.</p>
						<b>STEPS:</b>
						<ol>
							<li>Start <span class="small red awesome">Gaggle boss</span></li>
							<li>First we will collect regulon info from a website by adding a <span class="small awesome">firegoose</span> module. 
								Path to Firefox should be automatically detected. Leave subaction unchanged. Enter data url, or web resource to get regulon information.
								In this example we will use one of Methanococcus maripaludis GRN regulon. (http://baliga.systemsbiology.net/cmonkey/enigma/mmp/cmonkey_4.8.8_mmp_1661x58_11_Oct_11_16:14:07/htmls/cluster0114.html)
								</li>
							<li>Add <span class="small awesome">firegoose</span> module for DAVID.
								Select the "DAVID" option from subaction menu and connect to the first Firegoose module.
								<code>Manual: You need to convert the Namelist to a DAVID Id list once it has been submitted to DAVID.</code></li>
							<li>Add firegoose module for Entrez. Select "Entrez Protein" subaction and connect to the first Firegoose module.</li>
							<li>Last, we will add another firegoose module for STAMP. Select the "STAMP" subaction and connect to the first Firegoose module.</li>
								<code>Manual: You need to select an appropriate database for motif searching on the STAMP site.</code></li>
							<li> Optional: Save your workflow</li>
							<li> Run your workflow by using <span class="small green awesome">"Run"</span> button in the control panel.</li>	
						</ol>
					</div>
					<h4> Case 2 - Analyzing gene expression data</h4>
					<div> <img align="right" src="{% static "images/workflow_case-2.png" %}">
					<p> The second exercise includes both desktop applications and web resources. The aim of this workflow is to automate analysis of gene expression data and identify functional associations, functional enrichments and associated regulatory network modules. We will start by opening a gene expression file (tab delimited) in Mev. After performing statistical analyses, we will broadcast selected genes via Firegoose to KEGG (Metabolic pathway enrichment) and EMBL Strings (Functional associations). Inparallel, we will also open a Network model in Cytoscape and broadcast our gene list to the network in order to select network modules that include these genes.</p>   
						<b>STEPS:</b>
						<ol>
							<li>Start <span class="small red awesome">Gaggle boss</span></li>
							<li>We will add mev module to open a gene expression file. Leave subaction unchanged. If you are opening the gene expression file from the web, enter the url; otherwise, enter the path to the file on your computer.								In this example we will use gene expression data coming from studying energy metabolism for Desulfovibrio vulgaris.
								We will compare gene expression upon growth on Pyruvate+Sulfate versus Pyruvate/Fermentation.
								<code>Manual: After your workflow has started Mev will load the gene expression file. You can perform SAM analyses and then select genes that are upregulated in Pyruvate/Fermentation conditions and then broadcast them to the workflow.</code></li>
								</li>
							<li>Add firegoose module for EMBL Strings. Select "EMBL Strings" subaction and connect to MEV module. The EMBL Strings website will use the Namelist for our selected genes and search against the Strings database to identify functional associations. The resulting functional association network will be shown and captured in the workflow report.</li>
							<li>Add <span class="small awesome">firegoose</span> module for EMBL Strings
								Select "EMBL Strings" subaction and connect to MEV module. EMBL Strings website will use Namelist for our selected genes and search against Strings database 
								to identify functional associations. Resulting functional association network will be shown and captured in the workflow report.</li>
							<li>Last, we will add <span class="small awesome">Cytoscape</span> module to open a network file. (For this example we will use GRN model for D. vulgaris).
								Connect Cytoscape module to MEV module and enter path to your cytoscape network file (web or desktop). When workflow is started cytoscape will open 
								your network file and capture broadcast list from MEV. Upregulated genes will be selected in the network for further network manipulations.
								</li>
							<li> Optional: Save your workflow</li>
							<li> Run your workflow by using <span class="small green awesome">"Run"</span> button in the control panel.</li>	
						</ol>
					</div>
					<h4> Case 3 - Assigning functions to hypothetical proteins</h4>
					<div><img src="{% static "images/workflow_case-3.png" %}"></div>
					</div></p>
				</div>
		
            <div id="tabs-4"><h5> Network Portal Screencast</h5><br>
			<h2>Discovering the regulation of target genes</h2>
                 <iframe width="640" height="360" src="//www.youtube.com/embed/VNg0q70sLds?rel=0" frameborder="0" allowfullscreen></iframe>
                 <p>In this screencast, we will explore the regulation of hemA gene in Rhodobacter sphaeroides. hemA and hemT genes encode isozymes that catalyze the formation of 5-aminolevulinic acid, the first step in the biosynthesis of all tetrapyrroles present in Rhodobacter spaheroides 2.4.1.<br>
                 Previous studies have shown that hemA is transcribed from two promoters and that two global regulatory proteins, PrrA and FnrL, regulate hemA transcription in response to changes in oxygen tension.</br>
                 In this screencast, we will see if we can validate some of these results from our network model and identify novel regulatory influences.</p>

                <h2>Exploring gene function and regulation with Workspace</h2>
                <iframe width="640" height="360" src="//www.youtube.com/embed/X_OqDXdZUKU" frameborder="0" allowfullscreen></iframe>
                <p>In this screencast, we will learn how to explore genes, co-regulated genes, and regulators within a gene regulatory network. </p>
                <p>Here, we will use network information to learn more about one poorly characterized gene. Our biological question is to explore cytochrome gene function in Geobacter sulfurreducens.</p>
                <p>G. sulfurreducens can receive electrons from other microbial species, which is useful in microbial fuel cells. Electron transfer involves c-type cytochromes. However, the specialization of the >100 c-type cytochromes in the genomes is only beginning to be described.</p>

			</div>
			
			<div id="tabs-5"><h5> Get Involved</h5><br>
				<h2> Development </h2>
				<p>One of the key benefits of the Gaggle and Gaggle Workspace is that anyone can implement their own Geese. We hope to promote interoperability among bioinformatics applications by making this as easy as possible. Have an application you'd like to see included on this page? Please contact us by email (sturkars at systemsbiology.org) or through the discussion groups.</p>
				<p> We strongly encourage developers to take advantage of easy implementation features in Gaggle and Network Portal. If you are a developer who would like to integrate his/her desktop/web applications with Gaggle/Network Portal we provide all the support you need.</p>
				<h2>Gaggle Subversion Repository</h2>
				<p>
				The Gaggle's source code is all open source, licensed under the LGPL,
				and served using the
				<a href="http://subversion.tigris.org/" target="_blank" class="textlink">Subversion</a>
				source code control system.</p>
				<p>The current development branch of Gaggle is located in:
				<b>http://gaggle.systemsbiology.net/svn/gaggle/gaggle/trunk</b></p>
				<ul>
					<li> To use the Subversion archive hosted at the Institute for Systems Biology,
						take the following steps. (These instructions use unix-style commands; Windows
						users translate accordingly.)
						<ol>
							<li> Locate or install a Subversion client.  (Below we use the
								command-line client, <b><i>svn</i></b>.) </li>
							<li> <code> mkdir <i><b>yourInstallationDirectory</b></i> </code> </li>
							<li> <code> svn&nbsp;checkout&nbsp;http://gaggle.systemsbiology.net/svn/gaggle/gaggle/trunk&nbsp;<i><b>yourInstallationDirectory</b></i></code>
							</li>
						</ol>
					</li>
					<li> For authenticated access (allowing commits) to the Subversion archive, a password is required:
						<ul>
							<li> <code>svn&nbsp;checkout&nbsp;--username&nbsp;<i><b>username</b></i>&nbsp;--password&nbsp;<i><b>password</b></i>&nbsp;http://gaggle.systemsbiology.net/svn/gaggle/gaggle/trunk&nbsp;<i><b>yourInstallationDirectory</b></i></code></li>
						</ul>
					</li>
				</ul>
				
				
				<p>
				
				<h2>Firegoose Source code</h2>
				<ul>
					<li>Firegoose source is here:
						<a href="http://gaggle.systemsbiology.net/svn/gaggle/FireGoose/trunk/">http://gaggle.systemsbiology.net/svn/gaggle/FireGoose/trunk/</a></li>
				</ul>
				<h2>Network Portal Source code</h2>
				<ul>
					<li>Network Portal source is here:
						<a href="http://baliga-lab/github.com/network_portal">http://baliga-lab/github.com/network_portal</a></li>
				</ul>
				<h2>Gaggle Workspace API</h2>
				<ul>
					<li>Workspace API documentation is here:
						<a href="http://baliga-lab/github.com/network_portal">http://baliga-lab/github.com/network_portal</a></li>
				</ul>

			</div>       
       
	</div>

<!--
      <h5>Navigation</h5>
      <span class="search-help"> Mouse over to see details of specific navigation items</span>
      <img id="Image-Maps_9201202132102136" src="{% static "images/gene-page-map.png" %}" usemap="#Image-Maps_9201202132102136" border="0" width="724" height="620" alt="" />
      <map id="_Image-Maps_9201202132102136" name="Image-Maps_9201202132102136">
	<area shape="rect" coords="20,84,89,111" href="#" alt="<font color=red>HOME:</font><br> Takes you to the main Portal Page" title="home"    />

	<area shape="rect" coords="98,83,167,110" href="#" alt="<font color=red>SEARCH:</font><br> Main search page" title="search"    />
	
<area shape="rect" coords="178,85,255,112" href="#" alt="<font color=red>SPECIES:</font><br> List all the species currently available at the network portal with quick links to network information" title="species"    />

	<area shape="rect" coords="266,85,344,112" href="#" alt="<font color=red>FUNCTION:</font><br> Lists functional categories and relevant information from KEGG, GO, COG and TIGRFam." title="function"    />
	
	<area shape="rect" coords="358,84,433,111" href="#" alt="<font color=red>NETWORKS:</font><br> List of currently available networks" title="networks"    />
	
	<area shape="rect" coords="448,83,517,110" href="#" alt="<font color=red>HELP:</font><br> This help page" title="help"    />

	<area shape="rect" coords="17,122,87,149" href="#" alt="<font color=red>GENE NAME:</font><br>Locus tag for a given gene. Clicking will take you to the main summary page for that gene." title="gene-name"    />

	<area shape="rect" coords="89,122,159,149" href="#" alt="<font color=red>GENE NETWORK:</font><br>This view shows interactive <strong>Cytoscapeweb</strong> network for the gene if it belongs to any regulon.<br> Associated regulons, motifs and regulators are also included in the network." title="gene-network"    />

	<area shape="rect" coords="161,122,249,149" href="#" alt="<font color=red>GENE REGULATION:</font><br> This tab shows table view of regulatory influences on the gene. <br> If the gene is a regulator by itself, regulons that are regulated by it are also shown in another table." title="gene-regulation"    />

	<area shape="rect" coords="248,122,321,149" href="#" alt="<font color=red>GENE MOTIFS:</font><br> Motif tab will show motif logo and prediction statistics if any motif sequence is identified upstream of the gene." title="gene-motif"    />

	<area shape="rect" coords="323,122,396,149" href="#" alt="<font color=red>GENE FUNCTIONS:</font><br> You can find gene functional annotations from KEGG, COG, GO and TIGRFam here." title="gene-functions"    />

	<area shape="rect" coords="396,122,518,149" href="#" alt="<font color=red>GENE REGULON MEMBERS:</font><br> If gene is associated with a regulon(s), all other genes in these regulons are also listed in a table view." title="gene-regulon-members"    />

	<area shape="rect" coords="519,122,583,149" href="#" alt="<font color=red>GENE HELP:</font><br> Shows gene page specific help content" title="gene-help"    />

	<area shape="rect" coords="583,122,663,149" href="#" alt="<font color=red>GENE COMMENTS:</font><br> You can share your comments/questions/updates for a particular gene here by using network portal authentication or facebook connect. " title="gene-comments"    />

	<area shape="rect" coords="473,158,695,217" href="#" alt="<font color=red>GENE INFO:</font><br> This section of the main gene page gives short information about the gene such as organism, locus tag, gene name, aliases" title="gene-info"    />

	<area shape="rect" coords="141,210,224,247" href="#" alt="<font color=red>CIRCVIS PLASMID:</font><br> Plasmid section of the genome. <br>Each section of the circular visualisation that are divided by the black lines represent each chromosome/plasmid for the organism that this gene belongs to." title="circvis-plasmid"    />

	<area shape="rect" coords="181,432,264,469" href="#" alt="<font color=red>CIRCVIS CHROMOSOME:</font><br> Chromosome section of the genome. <br>Each section of the circular visualisation that are divided by the black lines represent each chromosome/plasmid for the organism that this gene belongs to." title="circvis-chromosome"    />

	<area shape="rect" coords="142,296,225,333" href="#" alt="<font color=red>CIRCVIS INTERACTIONS:</font><br> Red lines show connection of this gene to all other members of the regulons it belongs to. Mouseover the particular link to see details for the source and the target genes" title="circvis-interactions"    />

	<area shape="rect" coords="277,338,333,375" href="#" alt="<font color=red>CIRCVIS SOURCE GENE:</font><br> This is the source gene that is presented in this page (in this particular example DVU0946). <br> Mouseover to see more details." title="circvis-source-gene"    />

	<area shape="rect" coords="108,373,164,410" href="#" alt="<font color=red>CIRCVIS TARGET GENE:</font><br> All blue dots except the source gene represent target genes that share the same regulon as source gene. <br> Mouseover to see more details" title="circvis-target-gene"    />
	<area shape="rect" coords="311,211,367,268" href="#" alt="<font color=red>CIRCVIS LEGEND:</font><br> Legend for the Circular visualization" title="circvis-legend"    />

	<area shape="rect" coords="477,305,692,423" href="#" alt="<font color=red>GENE REGULATION:</font><br> This table gives quick statistics about the regulatory information for this gene including regulon numbers, # of regulators and number of conditions that are included in each regulon.<p> Clicking on the regulons and regulators will take you to GENE REGULATION tab to give more details." title="gene-regulation-table"    />

	<area shape="rect" coords="477,426,692,491" href="#" alt="<font color=red>GENE MOTIFS:</font><br> This table shows logos for predicted motifs and linked to GENE MOTIFS tab content for details." title="gene-motif-table"    />

	<area shape="rect" coords="47,496,454,595" href="#" alt="<font color=red>GENE FUNCTIONAL ANNOTATIONS:</font><br> This is the summary view for gene functional annotations and linked to GENE FUNCTIONS tab for details" title="gene-functional-annotations"    />
	<area shape="rect" coords="10,10,314,72" href="#" alt="portal-logo" title="portal-logo"    />
	<area shape="rect" coords="571,28,691,77" href="#" alt="isb-logo" title="isb-logo"    />
	<area shape="rect" coords="571,0,648,28" href="#" alt="about" title="about"    />
	<area shape="rect" coords="652,0,714,28" href="#" alt="contact" title="contact"    />
	<area shape="rect" coords="722,618,724,620" href="http://www.image-maps.com/index.php?aff=mapped_users_9201202132102136" alt="Image Map" title="Image Map" />
      </map>

      <h5>Coming soon...</h5>
      More detailed help pages are coming soon...-->
      
    </div><!-- End #main -->
    {% endblock %}
